airportsd.8
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "AIRPORTSD" "8" "" "User Manuals" ""
.hy
.SH NAME
.PP
airportsd - IATA airport code lookup server
.SH SYNOPSIS
.PP
airportsd [-v]
.SH DESCRIPTION
.PP
\f[I]airportsd\f[R] is an HTTP-accessible IATA airport code lookup
daemon.
We created this as a service students can use on lab machines for
developing an Ansible lookup plugin during trainings.
.PP
When running on a \f[I]tty\f[R] the program will output brief debugging
information on startup, and when it\[cq]s queried.
.PP
The (currently hard-coded) \f[I]CDB\f[R] database must exist and be
readable by the owner of the \f[I]airportsd\f[R] process.
.SH EXAMPLE
.IP
.nf
\f[C]
$ curl -i http://127.0.0.1:8812/lookup?iata=bcn
HTTP/1.1 200 OK
Date: Wed, 26 Mar 2024 22:17:20 GMT
Connection: close
Content-Type: application/json
Content-Length: 216
{
    \[dq]id\[dq]: \[dq]4004\[dq],
    \[dq]iata\[dq]: \[dq]BCN\[dq],
    \[dq]cc\[dq]: \[dq]ES\[dq],
    \[dq]city\[dq]: \[dq]Barcelona\[dq],
    \[dq]lat\[dq]: \[dq]41.2971\[dq],
    \[dq]lon\[dq]: \[dq]2.07846\[dq],
    \[dq]name\[dq]: \[dq]Josep Tarradellas Barcelona-El Prat Airport\[dq],
    \[dq]type\[dq]: \[dq]large_airport\[dq],
    \[dq]emoji\[dq]: \[dq]🇪🇸\[dq]
}
\f[R]
.fi
.SH ENDPOINTS
.PP
All \f[I]airportsd\f[R] API endpoints are obtained via GET requests.
.SS \f[V]lookup\f[R]
.PP
This endpoint expects \f[V]iata\f[R] query parameter with a case
insensitive 3-letter IATA airport code; the key is looked up in the
database and the JSON data or HTTP status code 404 are returned.
.SH OPTIONS
.PP
\f[I]airportsd\f[R] understands the following global options.
.TP
-v
show version information and exit
.SH ENVIRONMENT
.TP
\f[V]AIRPORTSD_IP\f[R]
optionally sets the listen address for \f[I]airportsd\f[R]; defaults to
\f[V]127.0.0.1\f[R] and we strongly recommend this is not changed to
anything other than a loopback address.
.TP
\f[V]AIRPORTSD_PORT\f[R]
optionally sets the TCP listen port to something other than the default
\f[V]8812\f[R].
.SH FILES
.TP
\f[V]/usr/local/share/airports.cdb\f[R]
the database file to query.
.SH REQUIREMENTS
.SS freebsd
.IP
.nf
\f[C]
$ pkg install libmicrohttpd tinycdb

$ cat > config.mk <<EOF
DBNAME= \[dq]/usr/local/share/airports.cdb\[dq]
LISTEN_HOST=   \[dq]127.0.0.1\[dq]
LISTEN_PORT=   \[dq]8812\[dq]
INC =           -I/usr/local/include
LIBS =          -L /usr/local/lib
EOF
\f[R]
.fi
.SS rhel/centos
.IP
.nf
\f[C]
yum install tinycdb
\f[R]
.fi
.SS debian
.IP
.nf
\f[C]
apt-get install tinycdb
\f[R]
.fi
.SS macos
.IP
.nf
\f[C]
brew install libmicrohttpd tinycdb
\f[R]
.fi
.SS all
.IP \[bu] 2
libmicrohttpd (https://www.gnu.org/software/libmicrohttpd/)
.IP \[bu] 2
tinycdb (http://www.corpit.ru/mjt/tinycdb.html)
.SH CREDITS
.IP \[bu] 2
Airport data gratefully put into the Public Domain by
ourairports.com (https://ourairports.com/data/).
.IP \[bu] 2
Country emoji flags from Emoji
Flags (https://github.com/jonathan-kosgei/emoji-flags)
.SH AVAILABILITY
.PP
<https://github.com/jpmens/airportsd>
.SH AUTHOR
.PP
Jan-Piet Mens <https://jpmens.net>


airportsd.c
/*
 * airportsd.c
 * Copyright (C) 2020 Jan-Piet Mens <jp@mens.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <signal.h>
#include <errno.h>
#include <time.h>
#include <ctype.h>
#include <getopt.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <cdb.h>
// #define MHD_PLATFORM_H
#include <microhttpd.h>
#include "version.h"

# if MHD_VERSION < 0x00097002
#   define MH_RETVAL int
# else
#   define MH_RETVAL enum MHD_Result
# endif

struct MHD_Daemon *mhdaemon;
struct cdb cdb;
bool tty = false;

static void catcher(int sig)
{
	if (sig == SIGINT) {
		MHD_stop_daemon(mhdaemon);
		exit(1);
	}
}

static void ignore_sigpipe()
{
	struct sigaction oldsig;
	struct sigaction sig;

	sig.sa_handler = &catcher;
	sigemptyset(&sig.sa_mask);
#ifdef SA_INTERRUPT
	sig.sa_flags = SA_INTERRUPT;	/* SunOS */
#else
	sig.sa_flags = SA_RESTART;
#endif
	if (0 != sigaction(SIGPIPE, &sig, &oldsig))
		fprintf(stderr,
		"Failed to install SIGPIPE handler: %s\n", strerror(errno));
}

static int send_content(struct MHD_Connection *conn, const char *page,
		const char *content_type, int status_code)
{
	int ret;
	struct MHD_Response *response;
	response = MHD_create_response_from_buffer(strlen(page),
				(void*) page, MHD_RESPMEM_MUST_COPY);
	if (!response)
		return (MHD_NO);
	MHD_add_response_header(response, "Content-Type", content_type);
	ret = MHD_queue_response(conn, status_code, response);
	MHD_destroy_response(response);
	return (ret);
}

static int send_page(struct MHD_Connection *conn, const char *page, int status_code)
{
	return send_content(conn, page, "text/plain", status_code);
}

static int get_lookup(struct MHD_Connection *connection)
{
	const char *iata;
	char *key, *kp, data[8192];
	unsigned keylen = 0, datalen;

	iata = MHD_lookup_connection_value(connection, MHD_GET_ARGUMENT_KIND, "iata");
	if (!iata || !*iata) {
		return send_page(connection, "Missing iata", 422);
	}

	if (!(key = strdup(iata))) {
		return send_page(connection, "out of memory", 422);
	}

	for (kp = key; kp && *kp; keylen++, kp++) {
		*kp = ( islower(*kp) ? toupper(*kp) : *kp );
	}

	if (tty) {
		fprintf(stderr, "Lookup: [%s]\n", key);
	}

	if (cdb_find(&cdb, key, keylen) > 0) {
		datalen = cdb_datalen(&cdb);

		/* cap length (and possibly break JSON content) */
		datalen = (datalen >= sizeof(data)) ? sizeof(data) - 1 : datalen;
		if (cdb_read(&cdb, data, datalen, cdb_datapos(&cdb)) == 0) {
			data[datalen++] = '\n';
			data[datalen] = 0;

			free(key);
			return send_content(connection, data, "application/json", MHD_HTTP_OK);
		}
	}
	free(key);
	return send_page(connection, "not found", 404);
}

MH_RETVAL handle_connection(void *cls, struct MHD_Connection *connection,
	const char *url, const char *method, const char *version,
	const char *upload_data, size_t *upload_data_size, void **con_cls)
{
	if (strcmp(method, "GET") != 0) {
		return (send_page(connection, "not allowed", MHD_HTTP_METHOD_NOT_ALLOWED));
	}

	if (strcmp(url, "/lookup") == 0) {
		return get_lookup(connection);
	}

	return send_page(connection, "four-oh-four", MHD_HTTP_NOT_FOUND);
}

int main(int argc, char **argv)
{
	struct sockaddr_in sad;
	char *s_ip, *s_port, *progname = *argv;
	unsigned short port;
	int ch, fd;

	tty = isatty(0);

	if ((s_ip = getenv("AIRPORTSD_IP")) == NULL)
		s_ip = LISTEN_HOST;
	if ((s_port = getenv("AIRPORTSD_PORT")) == NULL)
		s_port = LISTEN_PORT;
	port = atoi(s_port);

	while ((ch = getopt(argc, argv, "v")) != EOF) {
		switch (ch) {
			case 'v':
				printf("airportsd %s\n", VERSION);
				exit(0);
			default:
				fprintf(stderr, "Usage: %s [-v]\n", progname);
				exit(2);
		}
	}

	argc -= optind;
	argv += optind;

	if (tty) {
		fprintf(stderr, "%s: listening on %s:%d, db=%s\n", progname, s_ip, port, DBNAME);
	}

	memset(&sad, 0, sizeof (struct sockaddr_in));
	sad.sin_family = AF_INET;
	sad.sin_port = htons(port);
	if (inet_pton(AF_INET, s_ip, &(sad.sin_addr.s_addr)) != 1) {
		perror("Can't parse IP");
		exit(2);
	}


	if ((fd = open(DBNAME, O_RDONLY)) == -1) {
		perror(DBNAME);
		exit(3);
	}
	if (cdb_init(&cdb, fd) < 0) {
		fprintf(stderr, "Cannot cdb_init\n");
		exit(3);
	}

	ignore_sigpipe();
	signal(SIGINT, catcher);

	mhdaemon = MHD_start_daemon(MHD_USE_SELECT_INTERNALLY, 0, NULL, NULL,
		&handle_connection, NULL,
		MHD_OPTION_SOCK_ADDR, &sad,
		MHD_OPTION_END);

	if (mhdaemon == NULL) {
		perror("Cannot initialize HTTP daemon");
		return (1);
	}

	while (1) {
		sleep(60);
	}

	MHD_stop_daemon(mhdaemon);
	cdb_free(&cdb);
	close(fd);
	return (0);
}
